chain/chain/src/store/latest_witnesses.rs

```diff
@@ -0,0 +1,149 @@
+//! This module is responsible for storing the latest ChunkStateWitnesses observed
+//! by the node. The latest witnesses are stored in the database and can be fetched
+//! for analysis and debugging.
+//! The number of stored witnesses is limited. When the limit is reached
+//! the witness with the oldest height is removed from the database.
+//! At the moment this module is used only for debugging purposes.
+
+use near_primitives::stateless_validation::ChunkStateWitness;
+use near_store::DBCol;
+
+use crate::ChainStoreAccess;
+
+use super::ChainStore;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_store::db::LATEST_WITNESSES_INFO;
+use rand::rngs::OsRng;
+use rand::RngCore;
+
+/// Maximum size of the latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_SIZE: ByteSize = ByteSize::gb(4);
+
+/// Maximum size of a single latest witness stored in the database.
+const SINGLE_LATEST_WITNESS_MAX_SIZE: ByteSize = ByteSize::mb(128);
+
+/// Maximum number of latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_COUNT: u64 = 60 * 30;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct LatestWitnessesKey {
+    pub height: u64,
+    pub shard_id: u64,
+    /// Each witness has a random UUID to ensure that the key is unique.
+    /// It allows to store multiple witnesses with the same height and shard_id.
+    pub random_uuid: [u8; 16],
+}
+
+impl LatestWitnessesKey {
+    /// `LatestWitnessesKey` has custom serialization to ensure that the binary representation
+    /// starts with big-endan height and shard_id.
+    /// This allows to query using a key prefix to find all witnesses for a given height (and shard_id).
+    fn serialized(&self) -> [u8; 32] {
+        let mut result = [0u8; 32];
+        result[..8].copy_from_slice(&self.height.to_be_bytes());
+        result[8..16].copy_from_slice(&self.shard_id.to_be_bytes());
+        result[16..].copy_from_slice(&self.random_uuid);
+        result
+    }
+}
+
+/// Keeps aggregate information about all witnesses stored in `DBCol::LatestChunkStateWitnesses`.
+/// Used for enforcing limits on the number of witnesses stored in the database.
+#[derive(Debug, Clone, Copy, BorshSerialize, BorshDeserialize, PartialEq, Eq, Default)]
+struct LatestWitnessesInfo {
+    pub count: u64,
+    pub total_size: u64,
+}
+
+impl LatestWitnessesInfo {
+    pub fn is_within_limits(&self) -> bool {
+        self.count <= LATEST_WITNESSES_MAX_COUNT
```

Comment:
you check this before adding the new one, should <= be < then?

Reply:
The `info` represents the post-operation state, including the new witness, so I use `<` to make sure that this final state is within limits.

Reply:
I guess it could just as well be `<=`, one more byte/witness doesn't really matter that much.

Reply:
ok saw your other comment, thanks.